ksql-engine/src/test/resources/query-validation-tests/group-by.json

```diff
@@ -187,26 +713,159 @@
+      ],
+      "expectedException": {
+        "type": "io.confluent.ksql.util.KsqlStatementException",
+        "message": "GROUP BY does not support aggregate functions: SUM is an aggregate function."
+      }
+    },
+    {
+      "name": "aggregate function nested in arithmetic (stream->table)",
+      "statements": [
+        "CREATE STREAM TEST (d1 VARCHAR, d2 INT) WITH (kafka_topic='test_topic', value_format='DELIMITED');",
+        "CREATE TABLE OUTPUT AS SELECT COUNT(*) FROM TEST GROUP BY 1 + SUM(d2);"
       ],
       "expectedException": {
-        "type": "io.confluent.ksql.util.KsqlException",
-        "message": "Non-aggregate SELECT expression must be part of GROUP BY: [SUBSTRING(TEST.SOURCE, 0, 2)]"
+        "type": "io.confluent.ksql.util.KsqlStatementException",
+        "message": "GROUP BY does not support aggregate functions: SUM is an aggregate function."
       }
+    },
+    {
+      "name": "aggregate function nested in UDF (stream->table)",
+      "statements": [
+        "CREATE STREAM TEST (d1 VARCHAR, d2 INT) WITH (kafka_topic='test_topic', value_format='DELIMITED');",
+        "CREATE TABLE OUTPUT AS SELECT COUNT(*) FROM TEST GROUP BY SUBSTRING(d1, SUM(d2), 1);"
+      ],
+      "expectedException": {
+        "type": "io.confluent.ksql.util.KsqlStatementException",
+        "message": "GROUP BY does not support aggregate functions: SUM is an aggregate function."
+      }
+    },
+    {
+      "name": "with projection without aggregate functions (stream->table)",
+      "statements": [
+        "CREATE STREAM TEST (d1 VARCHAR, d2 INT) WITH (kafka_topic='test_topic', value_format='DELIMITED');",
+        "CREATE TABLE OUTPUT AS SELECT SUBSTRING(d2, 1, 2) FROM TEST GROUP BY d2;"
+      ],
+      "expectedException": {
+        "type": "io.confluent.ksql.util.KsqlStatementException",
+        "message": "GROUP BY requires columns using aggregate functions in SELECT clause."
+      }
+    },
+    {
+      "name": "UDAF nested in UDF in select expression (stream->table)",
+      "statements": [
+        "CREATE STREAM TEST (d1 VARCHAR) WITH (kafka_topic='test_topic', value_format='DELIMITED');",
+        "CREATE TABLE OUTPUT AS SELECT SUBSTRING('Mr Bugalicious', CAST(COUNT(*) AS INT), 1) FROM TEST GROUP BY d1;"
+      ],
+      "inputs": [
+        {"topic": "test_topic", "value": "x"},
+        {"topic": "test_topic", "value": "xxx"},
+        {"topic": "test_topic", "value": "y"},
+        {"topic": "test_topic", "value": "x"},
+        {"topic": "test_topic", "value": "xxx"}
+      ],
+      "outputs": [
+        {"topic": "OUTPUT", "key": "x", "value": "M"},
+        {"topic": "OUTPUT", "key": "xxx", "value": "M"},
+        {"topic": "OUTPUT", "key": "y", "value": "M"},
+        {"topic": "OUTPUT", "key": "x", "value": "r"},
+        {"topic": "OUTPUT", "key": "xxx", "value": "r"}
+      ]
+    },
+    {
+      "name": "UDAF nested in UDF in select expression (table->table)",
```

Comment:
Wow, these are so thorough! The only additional tests maybe worth adding I can think of are one that checks we handle inferred arguments correctly (e.g., `SELECT COUNT() GROUP BY f1`), and another to check that an aggregate select expression without a GROUP BY fails. Not critical since you've already got unit tests for these cases, though maybe at least the former is nice to have since we can verify that ROWTIME is used as the default argument.

Reply:
> one that checks we handle inferred arguments correctly (e.g., `SELECT COUNT() GROUP BY f1`)

Not sure I understand what you mean here... 'inferred arguments'... do mean `COUNT()` vs `COUNT(*)`??  If so, that such a test doesn't really belong in a `group-by` test, more a count specific one, or some general UDAF one, IMHO.

> aggregate select expression without a GROUP BY fails

Good call - I'll add one.

Reply:
> do mean `COUNT()` vs `COUNT(*)`
Yeah that's what I meant. Your point about such a test belonging in a test for UDAFs (or even the specific UDAF) is fair.

BTW do we still need the tests in `aggregate-table.json` after this PR? Looks like the tests you've added to `group-by.json` subsume those.

Reply:
Good call - hadn't spotted that.